### ISSUE TITLE
Provide a dialog to change the engine version of an existing container

### ISSUE THREAD

**Issue by madoar** [PhoenicisOrg/phoenicis#1832]
Currently it is not possible to change the engine version of an existing container from inside the container tab. Therefore I propose to provide a dialog which allows changing the engine version of an existing container.

### Reaching the dialog
The dialog can be reached via a new button on the overview tab in the container information panel.

### Content of the dialog
I propose to design the new dialog similar to the already existing engines tab. This means that the new dialog should consist of the following areas:
- a sidebar on the left side which contains a search box, a list of all available engine types (engine categories) and a list widget selector
- a `TabPane` for each engine subcategory of the selected engine type in the sidebar.
   Each tab in the `TabPane` contains a combined list widget with the corresponding engine versions
- a details panel which contains a confirmation panel to apply a selected engine version for the previously selected container

**Screenshot of the mentioned engines tab:**
![grafik](https://user-images.githubusercontent.com/18488086/52536719-8d299880-2d5e-11e9-8c45-9c0058e478f9.png)

### Changing the engine version of a container
The engine version change itself should be performed in a new `changeVersion` version, that should be added to the `Engine` interface. The `changeVersion` method could be defined as:

```java
boolean changeVersion(String containerName, Map<String, Object> version);
```

or if we want to define it analogous to `createContainer`:

```java
boolean changeVersion(String subCategory, String version, String containerName);
```

#### Implementing `changeVersion` for the wine engine
For the wine engine the following method needs to be implemented in https://github.com/PhoenicisOrg/scripts/blob/master/Engines/Wine/Engine/Implementation/script.js:

```javascript
changeVersion: function (containerName, subCategory, version) {
    // run wineserver -k
    this.install(subCategory, version);
    // update phoenicis.cfg
    // run wineserver -k
}
```

#### Implementing `changeVersion` for engine types that don't support version changes
For engine types that don't support version changes a dummy implementation needs to be added to their engine implementation. Such a dummy implementation could look like the following:

```javascript
function changeVersion(container, version) {
   // do nothing
}
```

or if we want to do provide an implementation which allows for a once only execution:

```javascript
function changeVersion(container, version) {
   if (!hasEngineInstalled(container)) {
      // install the engine
   }
}
```

### Selecting all allowed engine versions for a container
Containers are often only compatible with a subset of the engine versions. If you try to apply an incompatible engine version on a container, it is possible that the container breaks. To prevent this additionally a `selectCompatibleEngineVersions` method should be added for the engines. This method then returns a list of all compatible engine versions for a given container.

For more information see the discussion in https://github.com/PhoenicisOrg/phoenicis/pull/1829

**Comment by ImperatorS79** [comment]
For wine it would simply be:
```js
changeVersion: function (containerName, subCategory, version) {
    this.kill();
    this.install(subCategory, version);
	var parts = subCategory.split("-");
    var distribution = parts[0];
    var architecture = parts[2];

    var containerNameCleaned = containerName.replace(this._containerRegex, '');
    var containerDirectory = this._winePrefixesDirectory + "/" + containerNameCleaned + "/";

    var containerConfiguration = this._configFactory.open(containerDirectory + "/phoenicis.cfg");

    containerConfiguration.writeValue("wineVersion", version);
    containerConfiguration.writeValue("wineDistribution", distribution);
    containerConfiguration.writeValue("wineArchitecture", architecture);
	this.kill();
}
```

But I do not know how to do the GUI part ^^.

**Comment by plata** [comment]
Just thinking: we could also make
```
changeVersion: function (containerName)
```
and open a wizard from Javascript (like e.g. the custom installer script).

Maybe this would even be the cleaner solution because we could get rid of the `subcategory` and `version` which might not be applicable for other engines at all anyways.

**Comment by ImperatorS79** [comment]
Ok, how would you ask the user for the required wine version (There is the distribution plus the version that you can change) ?

**Comment by plata** [comment]
Not completely sure about the way the wizard is retrieved (might be required to create a new one) but yeah something in that direction.

**Comment by ImperatorS79** [comment]
How could I add a button for that in the container information tab ? (next to delete container for example)

**Comment by plata** [comment]
See https://github.com/PhoenicisOrg/phoenicis/blob/bd9af0cd71e4f7a2a111684603e51cd9ee8abe20/phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/container/skin/ContainerOverviewPanelSkin.java#L68

@madoar do you think this stuff would be easier to find if it was in https://github.com/PhoenicisOrg/phoenicis/tree/master/phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/containers ?

**Comment by ImperatorS79** [comment]
Okay, but how do I generate a "callback" that will call the javascript changeVersion method when the button is clicked ?

**Comment by madoar** [comment]
I'm still planning to remove the old `org.phoenicis.javafx.views.mainwindow` and `org.phoenicis.javafx.views.common` packages. I just don't have a lot of time to continue working on this currently.

About how to add a new button:
This is quite simple actually. Like @plata mentioned you need to add a new `Button` object to 

https://github.com/PhoenicisOrg/phoenicis/blob/bd9af0cd71e4f7a2a111684603e51cd9ee8abe20/phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/container/skin/ContainerOverviewPanelSkin.java#L77

This can be done similar to the other buttons, for example the delete button:

https://github.com/PhoenicisOrg/phoenicis/blob/bd9af0cd71e4f7a2a111684603e51cd9ee8abe20/phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/container/skin/ContainerOverviewPanelSkin.java#L69-L71

When implementing a new button you should be aware that the container overview has already been modularized by me. This means it has been split into three components:
- a view (skin) containing the actual visual components. This is the class you're adding the new `Button` object to
- a controller (behavior). This class doesn't exist for the container overview
- a model (control). This class contains all the internal fields the component knows of for the container overview this is https://github.com/PhoenicisOrg/phoenicis/blob/bd9af0cd71e4f7a2a111684603e51cd9ee8abe20/phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/container/control/ContainerOverviewPanel.java

To add behavior to your new `Button` object you need to add a callback method to the `onMouseClicked` listener (for an example see the other buttons inside the component).
Whenever the button has been clicked with the mouse the callback method is executed.

**Comment by ImperatorS79** [comment]
But how do I ask the callback to execute a javascript method ?

**Comment by plata** [comment]
Check where this is used: https://github.com/PhoenicisOrg/phoenicis/blob/8061ef25eda856405a56722bd5282641b1e890a7/phoenicis-engines/src/main/java/org/phoenicis/engines/EnginesManager.java#L64
The javascript engine object can be used like a Java object.

**Comment by madoar** [comment]
@ImperatorS79 To get an `EnginesManager` instance in the `ContainerOverviewPanel` you need to pass it through from `ContainerInformationPanel#containerEngineController`. For this purpose you need to do the following:

- add a getter method for `EnginesManager` to `ContainerEngineController`
- add a `private final ObjectProperty<ContainerEngineController> containerEngineController;` field to `ContainerOverviewPanel`, including an initialization command in the constructor and a property and getter method (see the other properties for an example)
- Pass the `containerEngineController` to the `ContainerOverviewPanel` in
   https://github.com/PhoenicisOrg/phoenicis/blob/bd9af0cd71e4f7a2a111684603e51cd9ee8abe20/phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/container/skin/ContainerInformationPanelSkin.java#L81-L94

Afterwards you should be able to access the `EngineManager` instance via your `getControl().getContainerEngineController().getEngineManager()` in `ContainerOverviewPanelSkin`.

If you have additional questions feel free to ask!

**Comment by ImperatorS79** [comment]
That's above my java skills ^^. I see a little bit what you mean (the general flow), but the implementation details are too difficult for me and I do not have time currently for such things.

**Comment by madoar** [comment]
@ImperatorS79 what do you require to continue? If you just require a button calling a method inside an interface, implemented by a script, I can implement that for you.

**Comment by ImperatorS79** [comment]
Just a button that call the `changeVersion `function in the container panel ^^.

**Comment by ImperatorS79** [comment]
The superseded issue was alpha-3.

**Comment by ImperatorS79** [comment]
Solved by #1970.
